README.md
# Document Understanding Solution

DUS leverages the power of Amazon Textract, Amazon Comprehend , Amazon Comprehend Medical Amazon OpenSearch Service and Amazon Kendra to provide digitization , domain-specific data discovery, redaction controls , structural component extraction and other document processing & understanding capabilities.

![img](./images/DUS_with_kendra.png)

## Architecture Diagram

![img](./images/DUS_Arch.png)

### Note

Current document formats supported : **PDF,JPG,PNG**

Current maximum document file size supported : **150MB**

Current concurrent document uploads (via UI ) supported : **100**

## 1. CICD Deploy

### Requirements

- aws cli

  `sudo yum -y install aws-cli`

- pip3 (Required to install packages)

  `curl https://bootstrap.pypa.io/get-pip.py -o get-pip.py && python3 get-pip.py`

### Getting Started with CICD Deploy

- Create a bucket to act as the target Amazon S3 distribution bucket

_Note:_ You will have to create an S3 bucket with the template 'my-bucket-name-<aws_region>'; aws_region is where you are testing the customized solution.

For example, you create a bucket called `my-solutions-bucket-us-east-1`,

- Now build the distributable:

```
chmod +x ./deployment/build-s3-dist.sh
./deployment/build-s3-dist.sh <bucket-name-minus-region> <solution-name> <version>
```

For example,

```
./deployment/build-s3-dist.sh my-solutions-bucket document-understanding-solution v1.0.0
```

- Deploy the distributable to an Amazon S3 bucket in your account. _Note:_ you must have the AWS Command Line Interface installed.

```
aws s3 cp ./deployment/global-s3-assets/ s3://my-bucket-name-<aws_region>/<solution_name>/<my-version>/ --recursive --acl bucket-owner-full-control --profile aws-cred-profile-name
aws s3 cp ./deployment/regional-s3-assets/ s3://my-bucket-name-<aws_region>/<solution_name>/<my-version>/ --recursive --acl bucket-owner-full-control --profile aws-cred-profile-name
```

- Get the link of the document-understanding-solution.template uploaded to your Amazon S3 bucket.
- Deploy the Document Understanding solution to your account by launching a new AWS CloudFormation stack using the link of the document-understanding-solution.template.
- If you wish to manually choose whether to enable Kendra or Read-only mode (default 'true' and 'false', respectively), you need to add `ParameterKey=KendraEnabled,ParameterValue=<true_or_false>` and `ParameterKey=ReadOnlyMode,ParameterValue=<true_or_false>` after the email parameter when calling `create-stack`.

```
aws cloudformation create-stack --stack-name DocumentUnderstandingSolutionCICD --template-url https://my-bucket-name-<aws_region>.s3.amazonaws.com/<solution_name>/<my_version>/document-understanding-solution.template --parameters ParameterKey=Email,ParameterValue=<my_email> --capabilities CAPABILITY_NAMED_IAM --disable-rollback
```

This solutions will create 7 S3 buckets that need to be manually deleted when the stack is destroyed (Cloudformation will only delete the solution specific CDK toolkit bucket. The rest are preserved to prevent accidental data loss).

- 2 for CICD
- 1 for solution specific CDK Toolkit
- 2/3 for documents ((sample and general documents and optionally 1 for Medical sample documents if opting for Amazon Kendra Integration)
- 1 for the client bucket
- 1 for access logs
- 1 for CDK toolkit (if this is the customer's first try with CDK)
- 1 for document bulk processing pipeline

The solution is set up to reserve lambda concurrency quota. This is both to limit the scale of concurrent Lambda invocations as well to ensure sufficient capacity is available for the smooth functioning of the demo. You can tweak the "API_CONCURRENT_REQUESTS" value in source/lib/cdk-textract-stack.ts for changing the concurrency Lambda limits

### Notes

- Do NOT change the `cicd` in package.json. This field is for the deployment system to use in CodePipeline
- Due to limitations of CodeCommit, you cannot use this deploy approach if you add a file to the solution that is above 6MB (for good measure, stay below 5MB)

## Development Deploy

The instructions below cover installation on Unix-based Operating systems like macOS and Linux.
You can use a AWS Cloud9 environment or EC2 instance (recommended: t3.large or higher on Amazon Linux platform) to deploy the solution

### Requirements

Please ensure you install all requirements before beginning the deployment

- aws cli

  `sudo yum -y install aws-cli`

- node 10+

  `sudo yum -y install nodejs`

- yarn

  `curl --silent --location https://dl.yarnpkg.com/rpm/yarn.repo | sudo tee /etc/yum.repos.d/yarn.repo`

  `sudo yum -y install yarn`

- tsc

  `npm install -g typescript`

- jq

  `sudo yum -y install jq`

- moto (Required for running the tests)
  `pip install moto==2.3.2`

- pip3 (Required to install packages)
  `curl https://bootstrap.pypa.io/get-pip.py -o get-pip.py && python3 get-pip.py`

To deploy using this approach, you must first set few values inside the `package.json` file in the `source` folder.

- Set your deployment region in the `stack->region` property, replacing `"%%REGION%%"`. This deployment will not pull the AWS region from your current AWS profile.

  _Note_ : The AWS services used in this solution are not all available in all AWS Regions. Supported regions include us-east-1,us-west-2,eu-west-1. Please refer the [AWS Regions Table](https://aws.amazon.com/about-aws/global-infrastructure/regional-product-services/) for the most up to date information on which regions support the all services in DUS are available.

- Enter your email into the `email` property, replacing `"%%USER_EMAIL%%"`
- If you want to use the [Classic mode](#classic-mode), set the enableKendra flag to `false`. For [Kendra-enabled mode](#kendra-enabled-mode), set the flag as `true`
- If you want to use the [Read-only (RO) mode](#read-only-mode), set the is isROMode flag to `true`.

Now switch to the source directory, and use yarn to deploy the solution:

```
cd ./source
```

```
yarn && yarn deploy
```

The cli will prompt for approval on IAM Roles and Permissions twice in the full deploy. Once for the backend stack and then again for the client stack. The cli will prompt for an email. After the deploy is complete, an email will be sent to address provided with credentials for logging in.

Note:

This will create 5 or 6 S3 buckets that will have to be manually deleted when the stack is destroyed (Cloudformation does not delete them, in order to avoid data loss).

- 2/3 for documents (sample and general documents and optionally 1 for Medical sample documents if opting for Amazon Kendra Integration)
- 1 for the client stack
- 1 for document bulk processing pipeline
- 1 for CDK toolkit (if this is your first time using CDK)

The solution is set up to reserve lambda concurrency quota. This is both to limit the scale of concurrent Lambda invocations as well to ensure sufficient capacity is available for the smooth functioning of the demo. You can tweak the "API_CONCURRENT_REQUESTS" value in source/lib/cdk-textract-stack.ts for changing the concurrency Lambda limits

### Development Deploy Commands

- `yarn deploy:backend` : deploys or updates the backend stack
- `yarn deploy:client` : deploys or updates the client app
- `yarn deploy:setup-samples` : push sample docs to s3
- `yarn deploy:setup-user` : initiated prompts to set up a user
- `yarn deploy:show` : displays the url of the client app
- `yarn destroy` : tears down the CloudFormation backend and client stacks

### Development Deploy Workflow and stack naming

The `package.json` script node `stackname` sets the stackname for the deploy commands. Throughout development it has been imperative to maintain multiple stacks in order to allow client app development and stack architecture development to work without creating breaking changes. When a new stackname is merged into develop it should have the most up to date deployments.

### Developing Locally

Once deployed into the AWS account, you can also deploy locally for web development
This application uses [next.js](https://github.com/zeit/next.js/) along with [next-scss](https://github.com/zeit/next-plugins/tree/master/packages/next-sass) — all documentation for those packages apply here. NOTE: This application uses the static export feature of next.js — be aware of the limited features available when using static export.

#### Start Dev Server

- Clone this repository
- Run `yarn` to install/update packages
- Run `yarn dev`
- Navigate to http://localhost:3000
- NOTE: The dev build is noticeably slower than the production build because pages are built/unbuilt on-demand. Also, the code in the dev build is uncompressed and includes extra code for debugging purposes.

#### Generate Production Build

- Run `yarn export` to create a static export of the application.
- In a terminal go to the `app/out` directory and run `python -m SimpleHTTPServer`
- Navigate to http://localhost:8000

### Code Quality Tools

This project uses [Prettier](https://prettier.io) to format code. It is recommended to install a [Prettier extension for your editor](https://prettier.io/docs/en/editors.html) and configure it to format on save. You can also run `yarn prettier` to auto-format all files in the project (make sure you do this on a clean working copy so you only commit formatting changes).

This project also uses ESLint and sass-lint to help find bugs and enforce code quality/consistency. Run `yarn lint:js` to run ESLint. Run `yarn lint:css` to run sass-lint. Run `yarn lint` to run them both.

### Generating License Report

Run `yarn license-report` to generate a license report for all npm packages. See output in `license-report.txt`.

## DUS Modes:

### Classic Mode

This is first release of the DUS solution. The major services included in this mode include Amazon OpenSearch Service, Amazon Textract, Amazon Comprehend and Amazon Comprehend Medical that allow digitization, information extraction and indexing in DUS.

### Kendra-Enabled Mode

In the Classic version, DUS supports searching/indexing of documents using Amazon OpenSearch Service
In the kendra enabled mode, Amazon Kendra is added as an additional capability and can be used for exploring features such as Semantic Search, Adding FAQs and Access Control Lists.
Simply set the ` enableKendra: "true"` in package.json
_Note:_ Amazon Kendra Developer edition is deployed as a part of this deployment.

### Read-Only Mode

In this mode, DUS will only be available in Read-Only mode and you will only be able to analyze the pre-loaded documents. You will not be able to upload documents from the web application UI. In order to enable the Read-Only mode, set ` isROMode: "true"` in package.json. By default, this mode is disabled.

## Notes

### Document Bulk Processing

DUS supports bulk processing of documents. During deploy, an S3 bucket for document bulk processing is created. To use the bulk processing mode, simply upload documents under the `documentDrop/` prefix.
In Kendra mode, you can also upload the corresponding access control list under `policy/` prefix in the same bucket with the following name convention _\<document-name>\.metadata.json_ Be sure to upload the access control policy first and then the document.

### Other

- To switch between the DUS Classic version and Amazon Kendra enabled version, please follow a fresh deploy (either in a different region/ deleting the stack) and avoid updating the CloudFormation stack for the existing mode. Currently, DUS does not have the feature to seamlessly switch between the 2 modes.
  More info available in this [issue](https://github.com/awslabs/document-understanding-solution/issues/83)
- Do NOT change the `cicd` in package.json. This field is for the deployment system to use in CodePipeline
- Due to limitations of CodeCommit, you cannot use this deploy approach if you add a file to the solution that is above 6MB (for good measure, stay below 5MB)

## Cost

- As you deploy this sample application, it creates different resources (Amazon S3 bucket, Amazon SQS Queue, Amazon DynamoDB table, OpenSearch Service (and potentially Amazon Kendra) cluster(s) and AWS Lambda functions etc.). When you analyze documents, it calls different APIs (Amazon Textract , Amazon Comprehend and Amazon Comprehend Medical) in your AWS account. You will get charged for all the API calls made as part of the analysis as well as any AWS resources created as part of the deployment. To avoid any recurring charges, delete stack using "yarn destroy".

- The CDK Toolkit stacks that are created during deploy of this solution are not destroyed when you tear down the solution stacks. If you want to remove these resources, delete the S3 bucket that contains `staging-bucket` in the name, and then delete the `CDKToolkit` stack.

- You are responsible for the cost of the AWS services used while running this reference
  deployment. The solution consists of some resources that have to be paid by the hour/size
  such as Amazon OpenSearch Service, Amazon Kendra and Amazon S3 while others are serverless technologies where
  costs are incurred depending on the number of requests.
  The approximate cost for the solution for 100 documents/day comes under $20/day for the Classic Mode and under $80/day for Kendra-Enabled Mode. For accurate and most up-to-date pricing information, refer [AWS Pricing](https://aws.amazon.com/pricing/)

## Delete demo application

1. CICD Deploy:

Either run `aws cloudformation delete-stack --stack-name {CICD stack}`, or go to Cloudformation in the AWS Console and delete the stack that ends with "CICD". You will also have to go to CodeCommit in the console and manually delete the Repository that was created during the deploy.

2. Development Deploy:

Make sure you are in the `source` directory, and then run `yarn destroy`.

## License

This project is licensed under the Apache-2.0 License.
You may not use this file except in compliance with the License. A copy of the License is located at
http://www.apache.org/licenses/

## Additional Notes

The intended use is for users to use this application as a reference architecture to build production ready systems for their use cases. Users will deploy this solution in their own AWS accounts and own the deployment, maintenance and updates of their applications based on this solution.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.14] - 2023-08-10
- Updated certifi and other dependencies 

## [1.0.13] - 2023-06-15

### Changed

- Node module updates
- Bug fix - address front-end document upload errors being experienced by some users

## [1.0.12] - 2023-06-07

### Changed

- Node and python library updates

## [1.0.11] - 2023-04-18

### Changed

- Node module updates
- Enabled Amazon S3 server access logging on logging bucket(s)

## [1.0.10] - 2023-03-06

### Added

- Support for GovCloud deployment. Note: because of the absence of Amazon CloudFront in GovCloud, the deployment would need
  setting up a webserver outside of the deployment and copying the client app (UI) to the webserver resource for the front-end
  to work.

### Fixed

- Bug fix - address AWS CodePipeline and hence deployment failure because of an incorrect option in the `aws logs` API call

## [1.0.9] - 2023-02-13

### Fixed

- AWS CodePipeline role permissions missing `sns:TagResource` causing the AWS CodeDeploy to fail

## [1.0.8] - 2023-01-09

### Changed

- Node module updates

## [1.0.7] - 2022-12-12

### Changed

- Node and python library updates
- remove unused files

## [1.0.6] - 2022-11-15

### Changed

- bug fix and update node modules

## [1.0.5] - 2022-09-28

### Changed

- update `next` node library to 12.2.4
- update `node-fetch` library to 3.2.10

## [1.0.4] - 2022-08-10

### Changed

- update `cdk` from 1.132 to 1.158.0
- update `aws-sdk` to 2.1182.0 for Lambdas with Nodejs runtimes
- update `boto3` to 1.24.37 for Lambdas with Python runtimes
- update `react` and `next` node libraries based on compatibility with Node 14.x runtime
- removed `next-sass` as newer version of `next` has the functionality built-in
- removed `node-sass` and added `sass` as the original library is deprecated (https://github.com/sass/node-sass)

## [1.0.3] - 2021-11-15

- Update `cdk` from 1.42.0 to 1.132.0

## [1.0.2] - 2021-04-29

- Update `next` from 10.0.3 to 10.0.4
- Update `aws-amplify` 2.1.0 to 3.3.25
- Remove `sass-lint`
- Resolve `netmask` and `y18n` vulnerabilities

## [1.0.1] - 2021-01-04

### Changed

- Updated the command to install the package moto in README and deployment scripts
- Updated the version of the package ini from 1.3.5 to 1.3.8
- Modified package.json to install version 0.21.1 for axios which is a dependency of amplify

## [1.0.0] - 2020-10-30

### Added

- Initial Release


lib/stack.js
const { Stack } = require('@aws-cdk/cdk')
const {
  AuthFlow,
  CfnIdentityPool,
  CfnUserPoolClient,
  CfnIdentityPoolRoleAttachment,
  SignInType,
  UserPool,
  UserPoolAttribute,
} = require('@aws-cdk/aws-cognito')
const {
  DynamoEventSource,
  SqsEventSource,
  SnsEventSource,
  S3EventSource,
} = require('@aws-cdk/aws-lambda-event-sources')
const apigateway = require('@aws-cdk/aws-apigateway')
const dynamodb = require('@aws-cdk/aws-dynamodb')
const es = require('@aws-cdk/aws-elasticsearch')
const events = require('@aws-cdk/aws-events')
const iam = require('@aws-cdk/aws-iam')
const lambda = require('@aws-cdk/aws-lambda')
const sns = require('@aws-cdk/aws-sns')
const sqs = require('@aws-cdk/aws-sqs')
const s3 = require('@aws-cdk/aws-s3')
// const s3deployment = require('@aws-cdk/aws-s3-deployment')

class TextractDemoStack extends Stack {
  constructor(scope, id, props) {
    super(scope, id, props)

    const resourceName = name => `${id}${name}`
    const addCorstoBucket = bucket => {
      const cfnBucket = bucket.node.findChild('Resource')
      cfnBucket.addPropertyOverride('CorsConfiguration', {
        CorsRules: [
          {
            AllowedOrigins: ['*'],
            AllowedMethods: ['HEAD', 'GET', 'PUT', 'POST', 'DELETE'],
            MaxAge: '3000',
            ExposedHeaders: ['ETag'],
            AllowedHeaders: ['*'],
          },
        ],
      })
    }
    //**********DocumentsS3Bucket*****************************
    const documentsS3Bucket = new s3.Bucket(this, resourceName('DocumentsS3Bucket'), {
      versioned: false,
    })
    addCorstoBucket(documentsS3Bucket)

    //**********SamplesS3Bucket*****************************
    const sampleS3Bucket = new s3.Bucket(this, resourceName('SamplesS3Bucket'), {
      versioned: false,
    })
    addCorstoBucket(sampleS3Bucket)

    //**********UserPool*****************************
    const userPool = new UserPool(this, resourceName('UserPool'), {
      autoVerifiedAttributes: [UserPoolAttribute.Email],
      poolName: resourceName('UserPool'),
      signInType: SignInType.Email,
    })
    //**********UserPoolClient*****************************
    const userPoolClient = new CfnUserPoolClient(this, resourceName('UserPoolClient'), {
      clientName: resourceName('UserPoolClient'),
      explicitAuthFlows: [AuthFlow.AdminNoSrp],
      generateSecret: false,
      userPoolId: userPool.userPoolId,
    })
    const identityPool = new CfnIdentityPool(this, resourceName('IdentityPool'), {
      identityPoolName: resourceName('IdentityPool'),
      allowUnauthenticatedIdentities: true,
      cognitoIdentityProviders: [
        {
          clientId: userPoolClient.userPoolClientId,
          providerName: userPool.userPoolProviderName,
        },
      ],
    })
    this.cognitoFederatedPrincipal = new iam.FederatedPrincipal('cognito-identity.amazonaws.com', {
      StringEquals: {
        'cognito-identity.amazonaws.com:aud': identityPool.identityPoolId,
      },
    })
    const authenticatedCognitoFederatedPrincipal = new iam.FederatedPrincipal(
      'cognito-identity.amazonaws.com',
      {
        StringEquals: {
          'cognito-identity.amazonaws.com:aud': identityPool.identityPoolId,
        },
        'ForAnyValue:StringLike': {
          'cognito-identity.amazonaws.com:amr': 'authenticated',
        },
      },
      'sts:AssumeRoleWithWebIdentity'
    )
    const unauthenticatedCognitoFederatedPrincipal = new iam.FederatedPrincipal(
      'cognito-identity.amazonaws.com',
      {
        StringEquals: {
          'cognito-identity.amazonaws.com:aud': identityPool.identityPoolId,
        },
        'ForAnyValue:StringLike': {
          'cognito-identity.amazonaws.com:amr': 'unauthenticated',
        },
      },
      'sts:AssumeRoleWithWebIdentity'
    )
    const authedRole = new iam.Role(this, 'AuthedRole', {
      roleName: resourceName('AuthedRole'),
      assumedBy: authenticatedCognitoFederatedPrincipal,
    })
    const unauthedRole = new iam.Role(this, 'UnauthedRole', {
      roleName: resourceName('UnauthedRole'),
      assumedBy: unauthenticatedCognitoFederatedPrincipal,
    })
    const authedPolicy = new iam.Policy(this, 'AuthedPolicy', {
      policyName: resourceName('AuthedRolePolicy'),
      statements: [
        new iam.PolicyStatement(iam.PolicyStatementEffect.Allow)
          .addAction('cognito-identity:*')
          .addAction('cognito-sync:*')
          .addAction('mobileanalytics:PutEvents')
          .addResource('*'),
        new iam.PolicyStatement(iam.PolicyStatementEffect.Allow)
          .addActions(['s3:*'])
          .addResources([documentsS3Bucket.bucketArn, `${documentsS3Bucket.bucketArn}/*`]),
        new iam.PolicyStatement(iam.PolicyStatementEffect.Allow)
          .addActions(['s3:getObject'])
          .addResource(`${sampleS3Bucket.bucketArn}/*`),
      ],
    })
    const unauthedPolicy = new iam.Policy(this, 'UnauthedPolicy', {
      policyName: resourceName('UnauthedRolePolicy'),
      statements: [
        new iam.PolicyStatement(iam.PolicyStatementEffect.Allow)
          .addAction('cognito-identity:*')
          .addAction('cognito-sync:*')
          .addAction('mobileanalytics:PutEvents')
          .addResource('*'),
      ],
    })
    authedPolicy.attachToRole(authedRole)
    unauthedPolicy.attachToRole(unauthedRole)
    this.identityPoolRoles = new CfnIdentityPoolRoleAttachment(this, 'IdentityPoolRoles', {
      identityPoolId: identityPool.identityPoolId,
      roles: {
        authenticated: authedRole.roleArn,
        unauthenticated: unauthedRole.roleArn,
      },
    })

    //**********Elasticsearch******************************
    const elasticSearch = new es.CfnDomain(this, resourceName('ElasticSearchCluster'), {
      elasticsearchVersion: '6.5',
      ebsOptions: {
        ebsEnabled: true,
        volumeSize: 30,
        volumeType: 'gp2',
      },
    })

    //**********SNS Topics******************************
    const jobCompletionTopic = new sns.Topic(this, resourceName('JobCompletion'))

    //**********IAM Roles******************************
    const textractServiceRole = new iam.Role(this, resourceName('TextractServiceRole'), {
      assumedBy: new iam.ServicePrincipal('textract.amazonaws.com'),
    })

    textractServiceRole.addToPolicy(
      new iam.PolicyStatement().addResource(jobCompletionTopic.topicArn).addAction('sns:Publish')
    )

    //**********DynamoDB Table*************************
    //DynamoDB table with links to output in S3
    const outputTable = new dynamodb.Table(this, resourceName('OutputTable'), {
      partitionKey: { name: 'documentId', type: dynamodb.AttributeType.String },
      sortKey: { name: 'outputType', type: dynamodb.AttributeType.String },
    })

    //DynamoDB table with links to output in S3
    const documentsTable = new dynamodb.Table(this, resourceName('DocumentsTable'), {
      partitionKey: { name: 'documentId', type: dynamodb.AttributeType.String },
      streamSpecification: dynamodb.StreamViewType.NewImage,
    })

    //**********SQS Queues*****************************
    //Input Queue for sync jobs
    const syncJobsQueue = new sqs.Queue(this, resourceName('SyncJobs'), {
      visibilityTimeoutSec: 30,
      retentionPeriodSec: 1209600,
    })

    //Input Queue for async jobs
    const asyncJobsQueue = new sqs.Queue(this, resourceName('AsyncJobs'), {
      visibilityTimeoutSec: 30,
      retentionPeriodSec: 1209600,
    })

    //Queue
    const jobResultsQueue = new sqs.Queue(this, resourceName('JobResults'), {
      visibilityTimeoutSec: 900,
      retentionPeriodSec: 1209600,
    })
    //Trigger
    jobCompletionTopic.subscribeQueue(jobResultsQueue)

    //**********Lambda Functions******************************

    // Helper Layer with helper functions
    const helperLayer = new lambda.LayerVersion(this, resourceName('HelperLayer'), {
      code: lambda.Code.asset('lambda/helper'),
      compatibleRuntimes: [lambda.Runtime.Python37],
      license: 'Apache-2.0',
      description: 'Helper layer.',
    })

    // Textractor helper layer
    const textractorLayer = new lambda.LayerVersion(this, resourceName('Textractor'), {
      code: lambda.Code.asset('lambda/textractor'),
      compatibleRuntimes: [lambda.Runtime.Python37],
      license: 'Apache-2.0',
      description: 'Textractor layer.',
    })

    // Boto3 layer
    const boto3Layer = new lambda.LayerVersion(this, resourceName('Boto3'), {
      code: lambda.Code.asset('lambda/boto3/boto3-layer.zip'),
      compatibleRuntimes: [lambda.Runtime.Python37],
      license: 'Apache-2.0',
      description: 'Boto3 layer.',
    })

    // Elastsearch layer
    const elasticSearchLayer = new lambda.LayerVersion(this, resourceName('ElasticSearchLayer'), {
      code: lambda.Code.asset('lambda/elasticsearch/es.zip'),
      compatibleRuntimes: [lambda.Runtime.Python37],
      license: 'Apache-2.0',
      description: 'Elasticsearch layer.',
    })

    //------------------------------------------------------------

    // Document processor (Router to Sync/Async Pipeline)
    const documentProcessor = new lambda.Function(this, resourceName('TaskProcessor'), {
      runtime: lambda.Runtime.Python37,
      code: lambda.Code.asset('lambda/documentprocessor'),
      handler: 'lambda_function.lambda_handler',
      timeout: 300,
      environment: {
        SYNC_QUEUE_URL: syncJobsQueue.queueUrl,
        ASYNC_QUEUE_URL: asyncJobsQueue.queueUrl,
      },
    })
    //Layer
    documentProcessor.addLayer(helperLayer)
    //Trigger
    documentProcessor.addEventSource(
      new DynamoEventSource(documentsTable, {
        startingPosition: lambda.StartingPosition.TrimHorizon,
        batchSize: 1,
      })
    )

    //Permissions
    documentsTable.grantReadWriteData(documentProcessor)
    syncJobsQueue.grantSendMessages(documentProcessor)
    asyncJobsQueue.grantSendMessages(documentProcessor)

    //------------------------------------------------------------
    // PDF Generator
    const pdfGenerator = new lambda.Function(this, 'PdfGenerator', {
      runtime: lambda.Runtime.Java8,
      code: lambda.Code.asset('lambda/pdfgenerator'),
      handler: 'DemoLambdaV2::handleRequest',
      memorySize: 3000,
      timeout: 900,
    })

    documentsS3Bucket.grantReadWrite(pdfGenerator)
    sampleS3Bucket.grantReadWrite(pdfGenerator)

    //------------------------------------------------------------

    // Sync Jobs Processor (Process jobs using sync APIs)
    const syncProcessor = new lambda.Function(this, resourceName('SyncProcessor'), {
      runtime: lambda.Runtime.Python37,
      code: lambda.Code.asset('lambda/syncprocessor'),
      handler: 'lambda_function.lambda_handler',
      reservedConcurrentExecutions: 1,
      timeout: 25,
      environment: {
        OUTPUT_BUCKET: documentsS3Bucket.bucketName,
        OUTPUT_TABLE: outputTable.tableName,
        DOCUMENTS_TABLE: documentsTable.tableName,
        ES_DOMAIN: elasticSearch.domainEndpoint,
        PDF_LAMBDA: pdfGenerator.functionName,
      },
    })

    //Layer
    syncProcessor.addLayer(helperLayer)
    syncProcessor.addLayer(textractorLayer)
    syncProcessor.addLayer(boto3Layer)
    syncProcessor.addLayer(elasticSearchLayer)

    //Trigger
    syncProcessor.addEventSource(
      new SqsEventSource(syncJobsQueue, {
        batchSize: 1,
      })
    )

    //Permissions
    documentsS3Bucket.grantReadWrite(syncProcessor)
    sampleS3Bucket.grantReadWrite(syncProcessor)
    outputTable.grantReadWriteData(syncProcessor)
    documentsTable.grantReadWriteData(syncProcessor)
    syncProcessor.addToRolePolicy(
      new iam.PolicyStatement().addAllResources().addActions('textract:*')
    )
    syncProcessor.addToRolePolicy(
      new iam.PolicyStatement()
        .addResource(`${elasticSearch.domainArn}/*`)
        .addActions('es:ESHttpGet', 'es:ESHttpPost', 'es:ESHttpPut')
    )

    //------------------------------------------------------------

    // Async Job Processor (Start jobs using Async APIs)
    const asyncProcessor = new lambda.Function(this, resourceName('ASyncProcessor'), {
      runtime: lambda.Runtime.Python37,
      code: lambda.Code.asset('lambda/asyncprocessor'),
      handler: 'lambda_function.lambda_handler',
      reservedConcurrentExecutions: 1,
      timeout: 15,
      environment: {
        ASYNC_QUEUE_URL: asyncJobsQueue.queueUrl,
        SNS_TOPIC_ARN: jobCompletionTopic.topicArn,
        SNS_ROLE_ARN: textractServiceRole.roleArn,
      },
    })

    //Layer
    asyncProcessor.addLayer(helperLayer)
    asyncProcessor.addLayer(boto3Layer)

    //Triggers
    // Run async job processor every minute
    //const rule = new events.EventRule(this, resourceName('Rule'), {
    //  scheduleExpression: 'rate(1 minute)',
    //})
    //rule.addTarget(asyncProcessor)

    // Run when a job is successfully complete
    //asyncProcessor.addEventSource(new SnsEventSource(jobCompletionTopic))

    asyncProcessor.addEventSource(
      new SqsEventSource(asyncJobsQueue, {
        batchSize: 1,
      })
    )

    // Permissions
    documentsS3Bucket.grantRead(asyncProcessor)
    sampleS3Bucket.grantRead(asyncProcessor)
    asyncJobsQueue.grantConsumeMessages(asyncProcessor)
    asyncProcessor.addToRolePolicy(
      new iam.PolicyStatement().addResource(textractServiceRole.roleArn).addAction('iam:PassRole')
    )
    asyncProcessor.addToRolePolicy(
      new iam.PolicyStatement().addAllResources().addAction('textract:*')
    )

    //------------------------------------------------------------

    // Async Jobs Results Processor
    const jobResultProcessor = new lambda.Function(this, resourceName('JobResultProcessor'), {
      runtime: lambda.Runtime.Python37,
      code: lambda.Code.asset('lambda/jobresultprocessor'),
      handler: 'lambda_function.lambda_handler',
      memorySize: 2000,
      reservedConcurrentExecutions: 50,
      timeout: 900,
      environment: {
        OUTPUT_BUCKET: documentsS3Bucket.bucketName,
        OUTPUT_TABLE: outputTable.tableName,
        DOCUMENTS_TABLE: documentsTable.tableName,
        ES_DOMAIN: elasticSearch.domainEndpoint,
        PDF_LAMBDA: pdfGenerator.functionName,
      },
    })

    // Layer
    jobResultProcessor.addLayer(helperLayer)
    jobResultProcessor.addLayer(textractorLayer)
    jobResultProcessor.addLayer(boto3Layer)
    jobResultProcessor.addLayer(elasticSearchLayer)

    // Triggers
    jobResultProcessor.addEventSource(
      new SqsEventSource(jobResultsQueue, {
        batchSize: 1,
      })
    )

    // Permissions
    outputTable.grantReadWriteData(jobResultProcessor)
    documentsTable.grantReadWriteData(jobResultProcessor)
    documentsS3Bucket.grantReadWrite(jobResultProcessor)
    sampleS3Bucket.grantReadWrite(jobResultProcessor)
    jobResultProcessor.addToRolePolicy(
      new iam.PolicyStatement().addAllResources().addAction('textract:*')
    )
    jobResultProcessor.addToRolePolicy(
      new iam.PolicyStatement().addResource(`${elasticSearch.domainArn}/*`).addAction('es:*')
    )

    //------------------------------------------------------------

    pdfGenerator.grantInvoke(syncProcessor)
    pdfGenerator.grantInvoke(jobResultProcessor)

    //------------------------------------------------------------

    // API Processor
    const apiProcessor = new lambda.Function(this, resourceName('ApiProcessor'), {
      runtime: lambda.Runtime.Python37,
      code: lambda.Code.asset('lambda/apiprocessor'),
      handler: 'lambda_function.lambda_handler',
      reservedConcurrentExecutions: 50,
      timeout: 60,
      environment: {
        CONTENT_BUCKET: documentsS3Bucket.bucketName,
        SAMPLE_BUCKET: sampleS3Bucket.bucketName,
        OUTPUT_TABLE: outputTable.tableName,
        DOCUMENTS_TABLE: documentsTable.tableName,
        ES_DOMAIN: elasticSearch.domainEndpoint,
      },
    })

    // Layer
    apiProcessor.addLayer(elasticSearchLayer)
    apiProcessor.addLayer(helperLayer)

    // Permissions
    documentsTable.grantReadWriteData(apiProcessor)
    documentsTable.grant(apiProcessor, ['dynamodb:DescribeTable'])
    outputTable.grantReadWriteData(apiProcessor)
    documentsS3Bucket.grantRead(apiProcessor)
    sampleS3Bucket.grantRead(apiProcessor)
    apiProcessor.addToRolePolicy(
      new iam.PolicyStatement().addResource(`${elasticSearch.domainArn}/*`).addAction('es:*')
    )
    apiProcessor.addToRolePolicy(
      new iam.PolicyStatement().addAllResources().addActions('comprehend:*')
    )

    //------------------------------------------------------------
    // API

    function addCorsOptionsAndMethods(apiResource, methods = []) {
      const options = apiResource.addMethod(
        'OPTIONS',
        new apigateway.MockIntegration({
          integrationResponses: [
            {
              statusCode: '200',
              responseParameters: {
                'method.response.header.Access-Control-Allow-Headers':
                  "'Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token,X-Amz-User-Agent'",
                'method.response.header.Access-Control-Allow-Origin': "'*'",
                'method.response.header.Access-Control-Allow-Credentials': "'false'",
                'method.response.header.Access-Control-Allow-Methods':
                  "'OPTIONS,GET,PUT,POST,DELETE'",
              },
            },
          ],
          passthroughBehavior: apigateway.PassthroughBehavior.Never,
          requestTemplates: {
            'application/json': '{"statusCode": 200}',
          },
        })
      )
      const methodResource = options.node.findChild('Resource')
      methodResource.propertyOverrides.methodResponses = [
        {
          statusCode: '200',
          responseModels: {
            'application/json': 'Empty',
          },
          responseParameters: {
            'method.response.header.Access-Control-Allow-Headers': true,
            'method.response.header.Access-Control-Allow-Methods': true,
            'method.response.header.Access-Control-Allow-Credentials': true,
            'method.response.header.Access-Control-Allow-Origin': true,
          },
        },
      ]
      methods.forEach(method => {
        apiResource.addMethod(method)
      })
    }

    const api = new apigateway.LambdaRestApi(this, resourceName('TextractAPI'), {
      handler: apiProcessor,
      proxy: false,
    })
    addCorsOptionsAndMethods(api.root)

    const searchResource = api.root.addResource('search')
    addCorsOptionsAndMethods(searchResource, ['GET'])

    const documentsResource = api.root.addResource('documents')
    addCorsOptionsAndMethods(documentsResource, ['GET'])

    const documentResource = api.root.addResource('document')
    addCorsOptionsAndMethods(documentResource, ['GET', 'POST', 'DELETE'])

    const redactResource = api.root.addResource('redact')
    addCorsOptionsAndMethods(redactResource, ['GET', 'POST'])
  }
}
module.exports = TextractDemoStack
